components/fxa-client/src/oauth.rs

```diff
@@ -363,6 +389,75 @@ impl FirefoxAccount {
     pub fn clear_access_token_cache(&mut self) {
         self.state.access_token_cache.clear();
     }
+
+    #[cfg(feature = "integration_test")]
+    pub fn new_logged_in(
+        config: crate::Config,
+        session_token: &str,
+        scoped_keys: HashMap<String, ScopedKey>,
+    ) -> Self {
+        let mut fxa = FirefoxAccount::with_config(config);
+        fxa.state.session_token = Some(session_token.to_owned());
+        scoped_keys.iter().for_each(|(key, val)| {
+            fxa.state.scoped_keys.insert(key.to_string(), val.clone());
+        });
+        fxa
+    }
+}
+
+pub struct AuthorizationPKCEParams {
+    pub code_challenge: String,
+    pub code_challenge_method: String,
+}
+
+pub struct AuthorizationParameters {
+    pub client_id: String,
+    pub scope: String,
+    pub state: String,
+    pub access_type: String,
+    pub pkce_params: Option<AuthorizationPKCEParams>,
+    pub keys_jwk: Option<String>,
+}
+
+// Not impl From<url::Url> because we could get an error
```

Comment:
Oh yeah, errors! I suggest implementing [`TryFrom`](https://doc.rust-lang.org/nightly/core/convert/trait.TryFrom.html) instead of leaving this comment.

Reply:
TIL about `TryFrom` 😃

Reply:
Honestly, it's my new favourite thing!